k2/csrc/ragged_ops.cu

```diff
@@ -1249,14 +1250,70 @@ std::vector<RaggedShape> GetPrefixes(RaggedShape &src,
 
 Ragged<int32_t> AddSuffixToRagged(Ragged<int32_t> &src,
                                   const Array1<int32_t> &suffix) {
-  K2_LOG(FATAL) << "Not Implemented!";
-  return Ragged<int32_t>();
+  NVTX_RANGE(K2_FUNC);
+  int32_t num_axes = src.NumAxes();
+  K2_CHECK_GE(num_axes, 2);
+  K2_CHECK_EQ(suffix.Dim(), src.TotSize(num_axes - 2));
+  ContextPtr &c = src.Context();
+  Array1<int32_t> dst_values(c, src.NumElements() + suffix.Dim());
+  RaggedShape dst_shape = ChangeSublistSize(src.shape, 1);
+  // "row_splits1" and "row_ids1" below are actually on the last axis. We name
+  // them with "1" so that we can use "idx01" and "idx0" for those indexes in
+  // lambda, following the naming convention explained in k2/csrc/utils.h
+  const int32_t *dst_row_splits1_data = dst_shape.RowSplits(num_axes - 1).Data(),
+                *dst_row_ids1_data = dst_shape.RowIds(num_axes - 1).Data(),
+                *src_values_data = src.values.Data(),
+                *suffix_data = suffix.Data();
+  int32_t *dst_values_data = dst_values.Data();
+
+  K2_EVAL(
+      c, dst_shape.NumElements(), lambda_copy_values, (int32_t idx01)->void {
+        int32_t idx0 = dst_row_ids1_data[idx01];
+        if (idx01 == dst_row_splits1_data[idx0 + 1] - 1) {
+          // idx01 points to the last element of this row; copy from suffix
+          dst_values_data[idx01] = suffix_data[idx0];
```

Comment:
Some elements in `suffix_data` will never be accessed
if some `idx0` has no corresponding idx01.

For example:
```
row splits: [ 0 2 2 3 3]
values: [x x y]
```
You will never get `idx0 == 1` inside the kernel in the current implementation.

Reply:
Not sure if your tests cover this kind of case.

Reply:
idx01 is an index into dst, so I think it's always has the corresponding idx0

Reply:
> idx01 is an index into dst, so I think it always has the corresponding idx0


Yes, the above statement is true. But the reverse is not.

Reply:
You can run `cu_ragged_test` locally and you will find that the random case will not pass.

Reply:
Sorry I don't quite catch the point. there are no empty sublists in dst, so each idx0 of dst will be obtained.

Reply:
> You can run `cu_ragged_test` locally and you will find that the random case will not pass.

For some reason (library conflicts after cmake), I am currently unable to build targets for tests.

Reply:
> there are no empty sublists in dst, so each idx0 of dst will be obtained.

You are correct if there are no empty sublists. But the current implementation does not handle
sublists that are empty.  The code does not check for empty sublists and `RandomRagged`
may return empty sublists. That is why all your passed tests contain non-empty lists.

Reply:
mm, I didn't see there's any problem of @freewym 's implementation, as there are no empty rows in `dest` ( we have called ChangedSublist to add 1 for each row/sublist).

Reply:
sorry, you're correct. The current code has been updated.